### PR TITLE
[Feature](bangc-ops): add box_iou_rotated op.

### PR DESCRIPTION
**Pull request by ZhangLearning**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

add new op box_iou_rotated

## 2. Modification

1) mlu-ops/bangc-ops/kernels/box_iou_rotated/*
2) mlu-ops/bangc-ops/mlu_op.h
3) mlu-ops/bangc-ops/mlu_op_kernel.h
5) mlu-ops/bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/*
6) mlu-ops/bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/box_iou_rotated/*
8) mlu-ops/docs/bangc-docs/user_guide/9_operators/*
9) mlu-ops/docs/bangc-docs/design_docs/box_iou_rotated/*

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](../docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [x] diff1, diff1 <= 3e-3
- [x] diff2, diff2 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.      |           Details            |      Check Results       |
|----------------|---------------------------|---------------------|
|        1       |          Supported hardware         | MLU270 <br> MLU290 <br>MLU370<br>MLU590|
|        2       |          Job types          |    block <br> U1 <br> U4    |
|        3       |         Layouts            |  NHWC 、NCHW、ARRAY etc    |
|        4       |         Whether multi-dimensions are supported              |                |
|        5       |          Whether element zero is supported             |               |
|        6       |         Data type(half/float)       |         half / float etc           |
|        7      |        Whether there is size limit           |             |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [x] Multidimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

| Test Point         | Acceptance Standard | Test Result (Error Message) |
| -------------- | -------- | -------------------- |
| Whether it conforms to the operator restriction | Normal error |                      |
| Whether illegal parameters are passed  | Normal error |                      |

### 3.2 Performance Test

See [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform ：MLU270

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

Platform ：MLU290

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

Platform：MLU370

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

### 3.3 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -3130,6 +3130,103 @@ mluOpStatus_t MLUOP_WIN_API mluOpVoxelPoolingForward(mluOpHandle_t handle,
...
+// Group:BoxIouRotated
+/*!
+ * @brief Computes the intersection-over-union (Jaccard index, IOU) of rotated
+ * bounding-boxes. If \b aligned is false, then calculate the IOUs
+ * between each rotated bounding-box of \b bbox1 and \b bbox2, otherwise
+ * the IOUs between each aligned pair of rotated bounding-box of \b bbox1
+ * and \b bbox2.
+ *
+ * @param[in] handle
+ * Input. Handle to an MLUOP context that is used to manage MLU devices and
```
Comment (tudejiang79) [inline bangc-ops/mlu_op.h:3142]: 把所有Input.去掉

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -3130,6 +3130,103 @@ mluOpStatus_t MLUOP_WIN_API mluOpVoxelPoolingForward(mluOpHandle_t handle,
...
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] bbox2
+ * Input. Pointer to the MLU memory that stores the input tensor \b bbox2.
+ * It has shape (m, 5), indicating (x, y, w, h, theta) for each row.
+ * Note that theta is in radian.
+ * @param[in] ious_desc
+ * Input. The descriptor of the output tensor. For detailed information,
+ * see ::mluOpTensorDescriptor_t.
+ * @param[out] ious
+ * Output. IOUs or IOFs of input rotated bounding-boxes. Pointer to the MLU
```
Comment (tudejiang79) [inline bangc-ops/mlu_op.h:3173]: Output.也去掉

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -3130,6 +3130,103 @@ mluOpStatus_t MLUOP_WIN_API mluOpVoxelPoolingForward(mluOpHandle_t handle,
...
+ * memory that stores the output tensor.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
+ *
+ * @par Formula
+ * - See "Box Iou Rotated Operator" section in "Cambricon BANGC OPS User Guide"
```
Comment (tudejiang79) [inline bangc-ops/mlu_op.h:3179]:
```suggestion

```

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/box_iou_rotated/box_iou_rotated.cpp [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#include <cmath>
+#include "box_iou_rotated.h"
+#include "mlu_op.h"
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/box_iou_rotated/box_iou_rotated.cpp:14]: 这个头文件可以不要吧

**Review by tudejiang79 — APPROVED** [review]

File: bangc-ops/kernels/box_iou_rotated/box_iou_rotated_union1.mlu [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+ * Copyright (C) [2019-2022] by Cambricon, Inc.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
```
Comment (tudejiang79) [inline bangc-ops/kernels/box_iou_rotated/box_iou_rotated_union1.mlu:11]: 这个版本说明错了

File: bangc-ops/kernels/box_iou_rotated/box_iou_rotated_nonaligned.h [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+ * Copyright (C) [2019-2022] by Cambricon, Inc.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
```
Comment (tudejiang79) [inline bangc-ops/kernels/box_iou_rotated/box_iou_rotated_nonaligned.h:11]: 版本说明错了

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -2295,22 +2295,22 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
  * @param[in] features
- * Pointer to the MLU memory that stores the features tensor. The shape of \b features 
+ * Pointer to the MLU memory that stores the features tensor. The shape of \b features
  * is [batch_num, H, W, C].
  * @param[in] rois_desc
  * The descriptor of rois tensor, containing dimension and the layout of rois.
  * For detailed information, see ::mluOpTensorDescriptor_t.
  * @param[in] rois
- * Pointer to the MLU memory that stores rois tensors. \b rois[i] consists of [batch_id, 
+ * Pointer to the MLU memory that stores rois tensors. \b rois[i] consists of [batch_id,
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:2313]: 这里的roi ，写成ROI tensors吧

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -2322,22 +2322,22 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
  * Pointer to the MLU memory that stores the output tensor.
- * 
+ *
  * @par Return
  * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM.
  *
  * @par Data Type
  * - This function supports the following data types for input tensor \b features, \b rois,
- *   and output tensor \b output. Data type of all tensors should be the same.
+ *   and output tensor \b output . Data type of all tensors should be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:2340]: Data types

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -3130,6 +3130,99 @@ mluOpStatus_t MLUOP_WIN_API mluOpVoxelPoolingForward(mluOpHandle_t handle,
                                                      const mluOpTensorDescriptor_t pos_memo_desc,
                                                      void *pos_memo);
 
+// Group:BoxIouRotated
+/*!
+ * @brief Computes the intersection-over-union (Jaccard index, IOU) of rotated
+ * bounding-boxes. If \b aligned is false, then calculate the IOUs
+ * between each rotated bounding-box of \b bbox1 and \b bbox2, otherwise
+ * the IOUs between each aligned pair of rotated bounding-box of \b bbox1
+ * and \b bbox2.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3139]: otherwise这句话没说完，是想表达什么意思呢？

File: bangc-ops/kernels/debug.h [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+// MLISA sreg `perf_cache_miss_num` is supported only on 1M.
+#if (__BANG_ARCH__ < 300)
+
+#define PERF_CACHE_MISS_BEGIN()                            \
+  uint32_t m_start = 0;                                    \
+  uint32_t m_end = 0;                                      \
+  __asm__ volatile(                                        \
+      "mv.sreg.gpr %%perf_read, 1;\n\t"                    \
+      "mv.gpr.sreg %[m_start], %%perf_cache_miss_num;\n\t" \
+      : [m_start] "+&r"(m_start))
```
Comment (DanieeelLiu) [inline bangc-ops/kernels/debug.h:93]: 这个是不是要封装，mlisa指令

File: bangc-ops/kernels/debug.h [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  do {                                                                  \
+    __asm__ volatile(                                                   \
+        "mv.sreg.gpr %%perf_read, 1;\n\t"                               \
+        "mv.gpr.sreg %[m_end], %%perf_cache_miss_num;\n\t"              \
+        : [m_end] "+&r"(m_end));                                        \
+    printf("[taskId : %d]", taskId);                                    \
+    printf(MLUGREEN " [%s:%d]" COLOUREND MLUBLUE " [MLULOG]:" COLOUREND \
+                    "Kernel Cache Miss: %u\n",                          \
+           __FILE__, __LINE__, m_end - m_start);                        \
+  } while (0)
```
Comment (DanieeelLiu) [inline bangc-ops/kernels/debug.h:105]: 同上

**Review by GraceGuanH — APPROVED** [review]

**Review by DanieeelLiu — APPROVED** [review]

**Review by WangYuan-cn — APPROVED** [review]

**Pull request merged**